WordPress/openverse-api
# Reuse images across jobs

## Description


This PR
- sets up a mechanism to reuse the images across jobs using GitHub Action artifacts
- uses multi-stage builds to drastically cut down the sizes of our Python images by a total of **82.07%** (see below)
- updates all images to Python 3.10
- removes `supervisord` from `ingestion_server` (fixes #425)
- unifies the `ingestion_server` and `indexer_worker` images 
- simplifies image build matrix configuration (conflict with #424)

To goal of the PR is to eventually unify with #424 in creating a comprehensive CI + CD pipeline that uses caching and artifacts to speedily test code, deploy images to GH CR and deploy docs to GH Pages. The presence of these artifacts can also make #213 easier.

## Image difference

| Name | Old | New |
|-|-|-|
| analytics | 1.24GB | 199MB | 
| ingestion_server | 1.12GB | 270MB |
| indexer_worker | 1.10GB | Gone! |
| openverse_api | 1.28GB | 401MB |
| Total | 4.74GB | 870MB |

## Testing Instructions

This PR makes changes to CI itself so all passing CI checks would be adequate. You can also see if all the services work just like before by following the steps in the `README.md`.

To see difference in image sizes,
- see current image sizes: `docker image ls`
- rebuild all images: `docker-compose build web ingestion_server analytics`
- see new image sizes: `docker image ls`
- compute, and optionally appreciate, the difference

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
